ktor-utils/jvm/src/io/ktor/util/cio/Semaphore.kt

```diff
@@ -51,12 +53,10 @@ class Semaphore(val limit: Int) {
      */
     fun leave() {
         var value = permits.incrementAndGet()
+        check(value <= limit)
 
         while (value > 0 && waiters.isNotEmpty()) {
-            val key = waiters.keys().nextElement() ?: continue
-            waiters.remove(key) ?: continue
-
-            key.resume(Unit)
+            waiters.poll()?.resume(Unit)
             value = permits.value
```

Comment:
The other issue is that it is likely that we always resume all since `resume()` and dispatch takes time so permits remain unchanged for some small amount of time that could be enough to poll and resume all the queue.